### DISCUSSION TITLE
Adding a custom user control to a view from a view model

### DISCUSSION THREAD

**Discussion by itslyn** [AvaloniaUI/Avalonia#15160] [Q&A] [answered]
Hello, I have a custom user control that I created which consists of a button with some text bound to it.

Here is the axaml:
![image](https://github.com/AvaloniaUI/Avalonia/assets/115854752/7c618027-ad96-4298-8c8c-9fab267f8960)

Here is its corresponding view model for the control:
![image](https://github.com/AvaloniaUI/Avalonia/assets/115854752/bc59e21d-c649-44af-856e-3d37582faaaf)

I have another parent View and ViewModel and I want to be able to dynamically add my custom control ActionBlock into the parent View using the parent's ViewModel. The amount of custom ActionBlocks I may need to add into this parent view is unknown and can change at any time. If I need to add 3 of my custom control, then I want them to stack sequentially.

Right now in the parent's view, I have many manually defined button controls and I want to replace these with dynamic ones of my new custom control.
![image](https://github.com/AvaloniaUI/Avalonia/assets/115854752/722bbc48-86e4-4dd2-b7bc-624abb388861)

How can I dynamically add custom user controls to a view from within a view model?
TIA!

**Comment by timunie** [comment]
Maybe ItemsControl with ItemTemplate beeing your UserControl?

**Reply by itslyn** [reply]
Thank you for your answer. Would you be able to explain a bit more on how I can add a dynamic amount of them from within a view model?

**Reply by thevortexcloud** [reply, marked as answer]
You make some sort of ItemsControl, then make an `ObservableCollection` of the view models you want to display. You bind your items control to the collection and set a data template for it (the template is what controls how the view model is presented). There are examples on the docs site.

https://docs.avaloniaui.net/docs/reference/controls/itemscontrol

Otherwise if you know the exact amount you need to display you can also use a `ContentControl` and bind each of them to a VM you want to display.

**Reply by itslyn** [reply]
Thank you so much! This was incredibly helpful. I have successfully gotten them to appear dynamically.

**Reply by thevortexcloud** [reply]
You are welcome. I would also suggest reading some WPF MVVM tutorials/doing the Avalonia Music store tutorial. It covers pretty much all of this and is a good way to learn XAML MVVM.

https://docs.avaloniaui.net/docs/tutorials/music-store-app/
